crates/store/src/db/models/queries/transactions.rs

```diff
@@ -126,14 +173,129 @@ pub struct TransactionSummaryRowInsert {
-            transaction_id: transaction_id.to_bytes(),
-            account_id: account_id.to_bytes(),
+            transaction_id: transaction_header.id().to_bytes(),
+            account_id: transaction_header.account_id().to_bytes(),
             block_num: block_num.to_raw_sql(),
+            initial_state_commitment: transaction_header.initial_state_commitment().to_bytes(),
+            final_state_commitment: transaction_header.final_state_commitment().to_bytes(),
+            input_notes: input_notes_binary,
+            output_notes: output_notes_binary,
+            size_in_bytes,
         }
     }
 }
+
+/// Select complete transaction headers for the given accounts and block range.
+///
+/// # Parameters
+/// * `account_ids`: List of account IDs to filter by
+///     - Limit: 0 <= size <= 1000
+/// * `block_range`: Range of blocks to include inclusive
+///
+/// # Returns
+/// A tuple of (`last_block_included`, `transaction_headers`) where:
+/// - `last_block_included`: The highest block number included in the response
+/// - `transaction_headers`: Vector of transaction headers, limited by payload size
+///
+/// # Note
+/// This function returns complete transaction header information including state commitments
+/// and note IDs, allowing for direct conversion to proto `TransactionHeader` without loading
+/// full block data. The response is size-limited to ~5MB to prevent excessive memory usage.
+/// If the limit is reached, transactions from the last block are excluded to maintain consistency.
+///
+/// # Raw SQL
+/// ```sql
+/// SELECT
+///     account_id,
+///     block_num,
+///     transaction_id,
+///     initial_state_commitment,
+///     final_state_commitment,
+///     input_notes,
+///     output_notes
+/// FROM transactions
+/// WHERE block_num > ?1 AND block_num <= ?2 AND account_id IN rarray(?3)
+/// ORDER BY block_num ASC
+/// LIMIT ?4
+/// ```
+pub fn select_transactions_headers(
+    conn: &mut SqliteConnection,
+    account_ids: &[AccountId],
+    block_range: RangeInclusive<BlockNumber>,
+) -> Result<(BlockNumber, Vec<crate::db::TransactionHeader>), DatabaseError> {
+    const MAX_PAYLOAD_BYTES: usize = 5 * 1024 * 1024; // 5 MB
+
+    QueryParamAccountIdLimit::check(account_ids.len())?;
+
+    if block_range.is_empty() {
+        return Err(DatabaseError::InvalidBlockRange {
+            from: *block_range.start(),
+            to: *block_range.end(),
+        });
+    }
+
+    let desired_account_ids = serialize_vec(account_ids);
+    let raw = SelectDsl::select(schema::transactions::table, TransactionHeaderRaw::as_select())
+        .filter(schema::transactions::block_num.gt(block_range.start().to_raw_sql()))
+        .filter(schema::transactions::block_num.le(block_range.end().to_raw_sql()))
+        .filter(schema::transactions::account_id.eq_any(desired_account_ids))
+        .order((schema::transactions::block_num.asc(),))
+        .load::<TransactionHeaderRaw>(conn)
+        .map_err(DatabaseError::from)?;
```

Comment:
This will work, but I'm concerned that it may return a lot of extra data and may become an attack vector. At the very least, we should put a `LIMIT` here. A simple way to do it would be to put something like 10,000 as the limit as we know 10,000 transactions are unlikely to fit into the request (assuming each one of them has at least 1 output note). This still may return too much data (because some transactions may contain multiple output notes).

A better approach may be to read data from the database in "chunks" - e.g., 1000 rows at a time (or maybe even fewer) and stop once the next chunk exceeds the total accumulated limit.

Reply:
Done!